### ISSUE TITLE
Why percious and recall suddenly increased ? 

### ISSUE THREAD

**Issue by GreenHandForCoding** [ultralytics/yolov5#10060]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I'm trying used yolov5 in product flaws, then some issues come.
The percious and recall suddenly increased in less 20 epoches.
I trained about 240 pictures, and I find percious and recall suddenly increased to about 0.995,  so why?
Detect results are unsatisfactory.

Maybe yolov5 doesn't fit this type of object. Should I modify parameters or expand the number of examples used in train? 

### Additional

![Uploading image.png…]()

**Comment by glenn-jocher** [comment]
👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by GreenHandForCoding** [comment]
Thanks!
Here is my another question:
I put some pictures of products without flaw  and the same quantity of empty labels  in training data, will it improve the performance of model?
I only used the defective products as my training data in the past.

**Comment by ExtReMLapin** [comment]
> Thanks! Here is my another question: I put some pictures of products without flaw and the same quantity of empty labels in training data, will it improve the performance of model? I only used the defective products as my training data in the past.

You mean add background images ? Images with no annotations to reduce false positives, yes, you can use like 10~15%.

I first used 70% and I had lower final score

**Comment by GreenHandForCoding** [comment]
![image](https://user-images.githubusercontent.com/106043395/200733600-341e7fbc-6c15-4f29-9ab7-48d15439f5d4.png)
![image](https://user-images.githubusercontent.com/106043395/200733645-10478ef2-028f-4418-8893-a98a680e76ee.png)
![image](https://user-images.githubusercontent.com/106043395/200734003-fafbdd31-cae0-44ce-be94-b14d5f934f6e.png)

Here are two graphs about the training process.
I think two graphs  indicate overfiting.  I trained about 600 pictures. Maybe too less training data or the characteristics of my data are not obvious  lead to overfiting.

Here is one.
![image](https://user-images.githubusercontent.com/106043395/200734689-99e5b197-292b-445a-9502-501c55591b02.png)

**Comment by ExtReMLapin** [comment]
What worries me about your metrics is the signal amplitude, precidion goes up and down a lot.


Could you show some annotated images (ones from the runs/exp00 folder for example ?

**Comment by freshche** [comment]
What are the parameters that you train with @GreenHandForCoding to achieve overfitting?

**Comment by GreenHandForCoding** [comment]
Hey! thank you and here are the two annotated images. @ExtReMLapin 

![train_batch0](https://user-images.githubusercontent.com/106043395/203249497-28d4845b-a00d-4364-8517-d524f95da7c0.jpg)
![train_batch1](https://user-images.githubusercontent.com/106043395/203249513-74117c65-79e8-4d81-8632-73c33d0f176d.jpg)

**Comment by ExtReMLapin** [comment]
What are your training settings ? Which model do you use and what are your images resolutions

**Comment by GreenHandForCoding** [comment]
Traning settings:
`def parse_opt(known=False):
    parser = argparse.ArgumentParser()
    parser.add_argument('--weights', type=str, default=ROOT / 'yolov5l.pt', help='initial weights path')
    parser.add_argument('--cfg', type=str, default='', help='model.yaml path')
    parser.add_argument('--data', type=str, default=ROOT / 'data/flaw.yaml', help='dataset.yaml path')
    parser.add_argument('--hyp', type=str, default=ROOT / 'data/hyps/hyp.scratch-low.yaml', help='hyperparameters path')
    parser.add_argument('--epochs', type=int, default=300, help='total training epochs')
    parser.add_argument('--batch-size', type=int, default=4, help='total batch size for all GPUs, -1 for autobatch')
    parser.add_argument('--imgsz', '--img', '--img-size', type=int, default=1280, help='train, val image size (pixels)')
    parser.add_argument('--rect', action='store_true', help='rectangular training')
    parser.add_argument('--resume', nargs='?', const=True, default=False, help='resume most recent training')
    parser.add_argument('--nosave', action='store_true', help='only save final checkpoint')
    parser.add_argument('--noval', action='store_true', help='only validate final epoch')
    parser.add_argument('--noautoanchor', action='store_true', help='disable AutoAnchor')
    parser.add_argument('--noplots', action='store_true', help='save no plot files')
    parser.add_argument('--evolve', type=int, nargs='?', const=300, help='evolve hyperparameters for x generations')
    parser.add_argument('--bucket', type=str, default='', help='gsutil bucket')
    parser.add_argument('--cache', type=str, nargs='?', const='ram', help='image --cache ram/disk')
    parser.add_argument('--image-weights', action='store_true', help='use weighted image selection for training')
    parser.add_argument('--device', default='', help='cuda device, i.e. 0 or 0,1,2,3 or cpu')
    parser.add_argument('--multi-scale', action='store_true', help='vary img-size +/- 50%%')
    parser.add_argument('--single-cls', action='store_true', help='train multi-class data as single-class')
    parser.add_argument('--optimizer', type=str, choices=['SGD', 'Adam', 'AdamW'], default='SGD', help='optimizer')
    parser.add_argument('--sync-bn', action='store_true', help='use SyncBatchNorm, only available in DDP mode')
    parser.add_argument('--workers', type=int, default=8, help='max dataloader workers (per RANK in DDP mode)')
    parser.add_argument('--project', default=ROOT / 'runs/train', help='save to project/name')
    parser.add_argument('--name', default='exp', help='save to project/name')
    parser.add_argument('--exist-ok', action='store_true', help='existing project/name ok, do not increment')
    parser.add_argument('--quad', action='store_true', help='quad dataloader')
    parser.add_argument('--cos-lr', action='store_true', help='cosine LR scheduler')
    parser.add_argument('--label-smoothing', type=float, default=0.0, help='Label smoothing epsilon')
    parser.add_argument('--patience', type=int, default=100, help='EarlyStopping patience (epochs without improvement)')
    parser.add_argument('--freeze', nargs='+', type=int, default=[0], help='Freeze layers: backbone=10, first3=0 1 2')
    parser.add_argument('--save-period', type=int, default=-1, help='Save checkpoint every x epochs (disabled if < 1)')
    parser.add_argument('--seed', type=int, default=0, help='Global training seed')
    parser.add_argument('--local_rank', type=int, default=-1, help='Automatic DDP Multi-GPU argument, do not modify')

    # Logger arguments
    parser.add_argument('--entity', default=None, help='Entity')
    parser.add_argument('--upload_dataset', nargs='?', const=True, default=False, help='Upload data, "val" option')
    parser.add_argument('--bbox_interval', type=int, default=-1, help='Set bounding-box image logging interval')
    parser.add_argument('--artifact_alias', type=str, default='latest', help='Version of dataset artifact to use')

    return parser.parse_known_args()[0] if known else parser.parse_args()`

@ExtReMLapin 
The image-size(1328*1168) is close to 1280, so 1280 is used.

**Comment by ExtReMLapin** [comment]
I was expecting your commandlines arguments, not python code but I guess the image resize is good enough.
Try using the P2 model instead of the default ones

https://github.com/ultralytics/yolov5/blob/master/models/hub/yolov5-p2.yaml

**Comment by GreenHandForCoding** [comment]
@ExtReMLapin what? code you have seen is the one I downloaded in the official website and it's python.
I'll try to use the p2 model. I will appreciate your suggestion

**Comment by ExtReMLapin** [comment]
Yes, I asked for YOUR training params, not to see the python code, I expected you to send us your command line you use to start the training

example : `python train.py --img 2048 --batch 10 --epochs 1500 --data test_data.yaml --weights " " --cfg yolov5n6.yaml --hyp hyp.scratch-low.yaml --patience 0`

Here you can see the model but there are others arguments like the image size, n of epochs etc

**Comment by ExtReMLapin** [comment]
To use th P2 model you would have to replace `yolov5n6.yaml` by `yolov5-p2.yaml`

**Comment by GreenHandForCoding** [comment]
@ExtReMLapin 
Here is information form console and it's in training. By the way do you have any advice on my later training ?
Such as : set rect = True ...

Console:
[34m[1mtrain: [0mweights=best_1280.pt, cfg=yolov5-p2.yaml, data=data\flaw.yaml, hyp=data\hyps\hyp.scratch-low.yaml, epochs=300, batch_size=4, imgsz=1280, rect=False, resume=False, nosave=False, noval=False, noautoanchor=False, noplots=False, evolve=None, bucket=, cache=None, image_weights=False, device=, multi_scale=False, single_cls=False, optimizer=SGD, sync_bn=False, workers=8, project=runs\train, name=exp, exist_ok=False, quad=False, cos_lr=False, label_smoothing=0.0, patience=100, freeze=[0], save_period=-1, seed=0, local_rank=-1, entity=None, upload_dataset=False, bbox_interval=-1, artifact_alias=latest
YOLOv5  2022-11-7 Python-3.9.7 torch-1.10.0+cu113 CUDA:0 (NVIDIA GeForce RTX 3060, 12287MiB)

[34m[1mhyperparameters: [0mlr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=0.05, cls=0.5, cls_pw=1.0, obj=1.0, obj_pw=1.0, iou_t=0.2, anchor_t=4.0, fl_gamma=0.0, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0
[34m[1mClearML: [0mrun 'pip install clearml' to automatically track, visualize and remotely train YOLOv5  in ClearML
[34m[1mComet: [0mrun 'pip install comet_ml' to automatically track and visualize YOLOv5  runs in Comet
[34m[1mTensorBoard: [0mStart with 'tensorboard --logdir runs\train', view at http://localhost:6006/
Overriding model.yaml nc=80 with nc=6

**Comment by ExtReMLapin** [comment]
The console logs itself won't really help, please setup wandb as described in the console to save the metrics, and then share the metrics with us, or directly the csv file in yolov5/runs/xxx/xxx

**Comment by GreenHandForCoding** [comment]
results:
![results](https://user-images.githubusercontent.com/106043395/205185889-1ff044a8-c822-4ecd-b818-a803843a0609.png)
P_curve:
![P_curve](https://user-images.githubusercontent.com/106043395/205185939-59ff551d-7025-4fd8-8a50-e59f96249b52.png)
PR_curve:
![PR_curve](https://user-images.githubusercontent.com/106043395/205185944-af31ab37-9be4-4879-aa8a-916cd0f9cd42.png)
R_curve:
![R_curve](https://user-images.githubusercontent.com/106043395/205185957-a1eed5bc-8103-4887-a206-bc2837fa9ca1.png)
train_batch0:
![train_batch0](https://user-images.githubusercontent.com/106043395/205186002-cbf53490-e6cc-4422-a47f-e491489e7431.jpg)
val_batch0:
![val_batch0_labels](https://user-images.githubusercontent.com/106043395/205186015-c774a48f-93c9-4b9c-ac06-fd225b94aceb.jpg)
@ExtReMLapin I've shown you some images. Tell me if you need more. Thank you.

**Comment by ExtReMLapin** [comment]
Alright, perfect, that's what we needed.
I would suggest top ignore the "bump" at the first iterations of the recall and precision, as mAP is still Low.

I would suggest going oogabooga mode and just increase the number of iterations from 300 here to 600 and see if mAP keeps going up.
As it keeps going up, you can "continue" training, here it's stopped a little soon I guess. Keep the selected settings, just increase the number of iteration, or triple/quadruple it if you have enough horsepower in your PC.

**Comment by GreenHandForCoding** [comment]
@ExtReMLapin  OK, understand. I will set 600  epochs and see what will happen. But the precious and recall decrease sharply from 0.8+ to 0.6- in the beginning makes me conscious.

**Comment by GreenHandForCoding** [comment]
@ExtReMLapin 
![results](https://user-images.githubusercontent.com/106043395/205774247-cfe49ef4-f806-48ae-959c-bdb4a49efd5f.png)
The result is not good...

**Comment by ExtReMLapin** [comment]
At the time you posted your answer, I had no idea, could you try to see if you didn't incorrectly annotate your images ?

Why are you using multiple classes ? I seems you're only trying to detect defects ? Can you try to make a copy of your dataset and change all the annotations to the same class ?

There are few defects you may have missed


![image](https://user-images.githubusercontent.com/3909752/210724830-50a77fdb-c280-4716-b06a-ea7fe59ae07f.png)

You annotate this : 

![image](https://user-images.githubusercontent.com/3909752/210724911-c9fb7499-e69a-41f7-85dd-04d95d3939b7.png)

But not theses ?

![image](https://user-images.githubusercontent.com/3909752/210724976-96bfc263-3c70-4114-bc7d-7b45858ac669.png)

![image](https://user-images.githubusercontent.com/3909752/210725075-19a65fc5-e55e-4eb1-8b1c-0fc34af2bc41.png)

**Comment by glenn-jocher** [comment]
@ExtReMLapin hi there! Thanks for sharing the results. It's possible that incorrect annotations or missed defects are affecting the performance. Given the nature of your task, you might consider simplifying your dataset to a single class and carefully reviewing your annotations. Additionally, it appears that some defects might have been missed during annotation. 
Feel free to dive deeper into your dataset and annotation, perhaps by creating a copy and ensuring all anomalies are appropriately annotated. This will likely lead to an improvement in your detection results. Good luck!
